[dist_optim] add distributed functional Adam optimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50620 [dist_optim] add distributed functional AdamW optimizer
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* #50618 [dist_optim] add distributed functional sgd optimizer
* **#45596 [dist_optim] add distributed functional Adam optimizer**
* #50411 [optimizer] refactor AdamW to use functional API
* #50410 [optimizer] refactor RMSProp to use functional API
* #50409 [optimizer] refactor Adadelta to use functional API
* #45597 [optimizer] refactor SGD to use functional API

Add TorchScript compatible Adam functional optimizer to distributed optimizer

wanchaol committed:
[dist_optim] add distributed functional Adam optimizer

Add TorchScript compatible Adam functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adam optimizer"


Add TorchScript compatible Adam functional optimizer to distributed optimizer

[ghstack-poisoned]

pritamdamania87 commented on torch/optim/functional.py:
```diff
@@ -73,8 +73,6 @@ def adam(params: List[Tensor],
         exp_avg = exp_avgs[i]
         exp_avg_sq = exp_avg_sqs[i]
         step = state_steps[i]
-        if amsgrad:
-            max_exp_avg_sq = max_exp_avg_sqs[i]
```
Was this a bug in the original implementation where amsgrad=False would fail? If so, do we have a unit test that covers this now?

wanchaol replied:
hmmm is not a bug, there're already test cases that test when `amsgrad=True` in [here](https://github.com/pytorch/pytorch/blob/939e0389def30aca9a799209a6d22f09e3b0583b/test/test_optim.py#L393). The reason I changed this is because this IF branch is merely just index the `max_exp_avg_sqs` when `amsgrad=True`, and use it in the below branch for `denom` calculation, but in TorchScript we don't support this dynamism, the variable has to be appear in both IF and ELSE branch in order for it to be used later.

pritamdamania87 replied:
Hmm maybe I'm missing something here, but if `amsgrad` was False, doesn't that mean that the variable `max_exp_avg_sq` is never actually defined and then used below?

wanchaol replied:
Sorry for late reply. Yes, if `amsgrad` was False, then `max_exp_avg_sq` never defined, and it's not used below since the`amsgrad` check also happens below. If it's `True`, then `max_exp_avg_sq` defined, and the below branch also can use it. 

This is due to python can have dynamic variable defined in one if branch and used if it potentially exist (although it might not exist, in that case python will dynamically throw error). The reason i changed this logic here is due to TorchScript couldn't handle such dynamism.

wanchaol committed:
Update on "[dist_optim] add distributed functional Adam optimizer"


Add TorchScript compatible Adam functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adam optimizer"


Add TorchScript compatible Adam functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adam optimizer"


Add TorchScript compatible Adam functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adam optimizer"


Add TorchScript compatible Adam functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol commented:
Seems something weird happened on this PR due to ghstack, and accidentally marked this PR as merged (though it's actually not...). Opening a new PR instead.
